serve.go

```diff
@@ -728,6 +731,64 @@ func debugHTTPmiddleware(next http.Handler) http.Handler {
 	})
 }
 
+func mdmAuthSignMessageMiddleware(db *boltdepot.Depot, next http.Handler) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		b64sig := r.Header.Get("Mdm-Signature")
+		if b64sig == "" {
+			http.Error(w, "Signature missing", http.StatusBadRequest)
+			return
+		}
+		sig, err := base64.StdEncoding.DecodeString(b64sig)
+		if err != nil {
+			http.Error(w, "Signature decoding error", http.StatusBadRequest)
+			return
+		}
+		p7, err := pkcs7.Parse(sig)
+		if err != nil {
+			http.Error(w, "Signature parsing error", http.StatusBadRequest)
+			return
+		}
+		bodyBuf, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			fmt.Println(err)
+			http.Error(w, "Problem reading request", http.StatusInternalServerError)
+			return
+		}
+
+		// the signed data is the HTTP body message
+		p7.Content = bodyBuf
+
+		// reassign body to our already-read buffer
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(bodyBuf))
+		// TODO: r.Body.Close() as we've ReadAll()'d it?
+
+		err = p7.Verify()
+		if err != nil {
+			http.Error(w, "Signature verification error", http.StatusBadRequest)
+			return
+		}
+
+		cert := p7.GetOnlySigner()
+		if cert == nil {
+			http.Error(w, "Invalid signer", http.StatusBadRequest)
+			return
+		}
+
+		if hasCN, err := HasCN(db, cert.Subject.CommonName, 0, cert, false); err != nil {
+			fmt.Println(err)
+			http.Error(w, "Unable to validate signature", http.StatusInternalServerError)
+		} else if hasCN != true {
```

Comment:
Good catch on the return; missed that!

I like the else there — where I can I like to keep variables scoped as small/short as possible and the bigger else statement achieves that. But I don't care enough to keep it that way. :) Yes, definitely `!hasCN`. That's what I get for late night coding!